# Optimization - Homework 2

Submitted by: Ramya Madhuri Desineedi (rd32895)

Date of Submission: 24th September 2021

# Question 1 - Max's Eating Contest

In [1]:
!pip install gurobipy

In [2]:
import numpy as np
import matplotlib.pyplot as plt
import gurobipy as gp

In [3]:
# Taking #tortes and #pies as x and y
# 2x + 3y <= 60
# f(x,y) = 4x + 5y -> Objective: To maximise points
# x, y >= 0 
obj = np.array([4,5]) # Objective
A = np.zeros((1,2)) # Initializing constraint matrix
A[0:] = [2,3] # Time constraint
B = np.array([60])
sense = np.array(['<'])

In [4]:
ojModel = gp.Model() # initialize an empty model

ojModX = ojModel.addMVar(2) # tell the model how many variables there are
# must define the variables before adding constraints because variables go into the constraints
ojModCon = ojModel.addMConstrs(A, ojModX, sense, B) # add the constraints to the model
ojModel.setMObjective(None,obj,0,sense=gp.GRB.MAXIMIZE) 
ojModel.Params.OutputFlag = 0

Academic license - for non-commercial use only - expires 2022-09-03
Using license file /Users/ramya/gurobi.lic


In [5]:
ojModel.optimize() # solve the LP

In [6]:
ojModel.objVal # Max's Optimal score

120.0

In [7]:
ojModX.x # how many tortes and pies to eat

array([30.,  0.])

## Applying additional constraints for Question 1

In [8]:
# Taking #tortes and #pies as x and y
# 2x + 3y <= 60
# f(x,y) = 4x + 5y -> Objective: To maximise points
# x, y >= 0 
#New Constraint
# y >= x or -x + y >=0
obj = np.array([4,5])
A = np.zeros((2,2))
A[0:] = [2,3] # time constraint
A[1:] = [-1,1] # pie constraint
B = np.array([60, 0])
sense = np.array(['<', '>'])

In [9]:
ojModel = gp.Model() # initialize an empty model

ojModX = ojModel.addMVar(2) # tell the model how many variables there are
# must define the variables before adding constraints because variables go into the constraints
ojModCon = ojModel.addMConstrs(A, ojModX, sense, B) # add the constraints to the model
ojModel.setMObjective(None,obj,0,sense=gp.GRB.MAXIMIZE) 
ojModel.Params.OutputFlag = 0

In [10]:
ojModel.optimize() # solve the LP

In [11]:
ojModel.objVal # Max's Optimal score

108.0

In [12]:
ojModX.x # how many tortes and pies to eat

array([12., 12.])

In [13]:
# By how many points does this constraint decrease Max’s total points? 
120 - 108

12

# Question 2 - Farmer

In [14]:
# w, c acres of wheat and corn respectively
# w + c <= 450
# to maximise profit in k$: 2w + 3c
# 3w + 2c <= 1000
# 2w + 4c <= 1200
obj = np.array([2,3])
A = np.zeros((3,2))
A[0:] = [1,1] # field size constraint
A[1:] = [3,2] # labour constraint
A[2:] = [2,4] # fertiliser constraint
B = np.array([450, 1000, 1200])
sense = np.array(['<', '<', '<'])

In [15]:
ojModel = gp.Model() # initialize an empty model

ojModX = ojModel.addMVar(2) # tell the model how many variables there are
# must define the variables before adding constraints because variables go into the constraints
ojModCon = ojModel.addMConstrs(A, ojModX, sense, B) # add the constraints to the model
ojModel.setMObjective(None,obj,0,sense=gp.GRB.MAXIMIZE) 
ojModel.Params.OutputFlag = 0

In [16]:
ojModel.optimize() # solve the LP
ojModel.objVal # Optimal profit

1000.0

In [17]:
ojModX.x # how many acres of wheat and corn

array([200., 200.])

In [18]:
fertiliser_available = list(range(200, 2201, 100))
fertiliser_available
for fert in range(len(fertiliser_available)):
    B[2] = fertiliser_available[fert]
    ojModel = gp.Model() # initialize an empty model
    ojModX = ojModel.addMVar(2) # tell the model how many variables there are
    # must define the variables before adding constraints because variables go into the constraints
    ojModCon = ojModel.addMConstrs(A, ojModX, sense, B) # add the constraints to the model
    ojModel.setMObjective(None,obj,0,sense=gp.GRB.MAXIMIZE) 
    ojModel.Params.OutputFlag = 0
    ojModel.optimize() # solve the LP
    print(fertiliser_available[fert], 'tons of fertiliser for', ojModX.x[0], 'acres of wheat')

200 tons of fertiliser for 100.0 acres of wheat
300 tons of fertiliser for 150.0 acres of wheat
400 tons of fertiliser for 200.0 acres of wheat
500 tons of fertiliser for 250.0 acres of wheat
600 tons of fertiliser for 300.0 acres of wheat
700 tons of fertiliser for 325.0 acres of wheat
800 tons of fertiliser for 300.0 acres of wheat
900 tons of fertiliser for 275.0 acres of wheat
1000 tons of fertiliser for 250.0 acres of wheat
1100 tons of fertiliser for 225.0 acres of wheat
1200 tons of fertiliser for 200.0 acres of wheat
1300 tons of fertiliser for 175.0 acres of wheat
1400 tons of fertiliser for 150.0 acres of wheat
1500 tons of fertiliser for 125.0 acres of wheat
1600 tons of fertiliser for 100.0 acres of wheat
1700 tons of fertiliser for 50.0 acres of wheat
1800 tons of fertiliser for 0.0 acres of wheat
1900 tons of fertiliser for 0.0 acres of wheat
2000 tons of fertiliser for 0.0 acres of wheat
2100 tons of fertiliser for 0.0 acres of wheat
2200 tons of fertiliser for 0.0 acres

From above results, Farmer discontinue producing wheat at 1800 tons of fertiliser

# Question 3 - Star Oil Company

In [19]:
# Assuming w1, w2, w3, w4, w5 are the fractions of investments 1, 2, 3, 4, 5 
# Hence, 0 <= w1, w2, w3, w4, w5 <= 1
# Objective is to maximise NPV: 13w1 + 16w2 + 16w3 + 14w4 + 39w5
# 11w1 + 53w2 + 5w3 + 5w4 + 29w5 <= 40
# 3w1 + 6w2 + 5w3 + 1w4 + 34w5 <=20
obj = np.array([13, 16, 16, 14, 39])
A = np.zeros((7,5))
A[0:] = [1,0,0,0,0]
A[1:] = [0,1,0,0,0]
A[2:] = [0,0,1,0,0]
A[3:] = [0,0,0,1,0]
A[4:] = [0,0,0,0,1]
A[5:] = [11,53,5,5,29] 
A[6:] = [3,6,5,1,34] 
B = np.array([1,1,1,1,1,40,20])
sense = np.array(['<', '<', '<', '<', '<', '<', '<'])

In [20]:
ojModel = gp.Model() # initialize an empty model

ojModX = ojModel.addMVar(5) # tell the model how many variables there are
# must define the variables before adding constraints because variables go into the constraints
ojModCon = ojModel.addMConstrs(A, ojModX, sense, B) # add the constraints to the model
ojModel.setMObjective(None,obj,0,sense=gp.GRB.MAXIMIZE) 
ojModel.Params.OutputFlag = 0

In [21]:
ojModel.optimize() # solve the LP
ojModX.x[2]

1.0

Star Oil should invest 100% of funds in investment 3

# Question 4 - Daily Nutrition

In [22]:
# c, m, b are servings of corn, milk and bread
# to minimise cost: 0.18c + 0.23m + 0.05b
# subject to: 0 <= c, m, b <= 10 -> servings limitation
# 2000 <= 72c + 121m + 65b <= 2250 -> calorie restriction
# 5000 <= 107c + 500m + 0b <= 50000 -> vitamin restriction
obj = np.array([0.18,0.23,0.05])
A = np.zeros((7,3))
# Amount of serving constraint
A[0:] = [1,0,0] 
A[1:] = [0,1,0] 
A[2:] = [0,0,1] 
# calorie constraint
A[3:] = [72,121,65] 
A[4:] = [72,121,65] 
# vitamin constraint
A[5:] = [107,500,0] 
A[6:] = [107,500,0] 
B = np.array([10,10,10,2000,2250,5000,50000])
sense = np.array(['<','<','<', '>', '<', '>', '<'])

In [23]:
ojModel = gp.Model() # initialize an empty model

ojModX = ojModel.addMVar(3) # tell the model how many variables there are
# must define the variables before adding constraints because variables go into the constraints
ojModCon = ojModel.addMConstrs(A, ojModX, sense, B) # add the constraints to the model
ojModel.setMObjective(None,obj,0,sense=gp.GRB.MINIMIZE) 
ojModel.Params.OutputFlag = 0

In [24]:
ojModel.optimize() # solve the LP
ojModX.x[0]

1.9444444444444444

I should be eating 1.94 servings of corn

# Question 5 - Paper and Wood Products

In [25]:
# let a1,a2,a3 be acres of forest cut in unit 1 in year 1, 2, 3
# let b1,b2,b3 be acres of forest cut in unit 2 in year 1, 2, 3
# Objective is to maximise : a1 + 1.3a2 + 1.4a3 + b1 + 1.2b2 + 1.6b3

#Constraints are listed as below:
# Total Acreage
# a1 + a2 + a3 <= 2
# b1 + b2 + b3 <= 3

# a1 + b1 >= 1.2
# a1 + b1 <= 2
# 1.3a2 + 1.2b2 >= 1.5

# 1.3a2 + 1.2b2 <= 2
# 1.4a3 +1.6b3 >= 2
# 1.4a3 +1.6b3 <= 3

obj = np.array([1,1.3,1.4,1,1.2,1.6])
A = np.zeros((8,6))
A[0:] = [1,1,1,0,0,0] # unit 1 size constraint
A[1:] = [0,0,0,1,1,1] # unit 2 size constraint
A[2:] = [1,0,0,1,0,0] # year 1 economic constraint
A[3:] = [1,0,0,1,0,0] # year 1 equipment and personnel constraint
A[4:] = [0,1.3,0,0,1.2,0] # year 2 economic constraint
A[5:] = [0,1.3,0,0,1.2,0] # year 2 equipment and personnel constraint
A[6:] = [0,0,1.4,0,0,1.6] # year 3 economic constraint
A[7:] = [0,0,1.4,0,0,1.6] # year 3 equipment and personnel constraint
B = np.array([2,3,1.2,2,1.5,2,2,3])
sense = np.array(['<', '<', '>', '<', '>', '<', '>', '<'])

In [26]:
ojModel = gp.Model() # initialize an empty model

ojModX = ojModel.addMVar(6) # tell the model how many variables there are
# must define the variables before adding constraints because variables go into the constraints
ojModCon = ojModel.addMConstrs(A, ojModX, sense, B) # add the constraints to the model
ojModel.setMObjective(None,obj,0,sense=gp.GRB.MAXIMIZE) 
ojModel.Params.OutputFlag = 0

In [27]:
ojModel.optimize() # solve the LP
round(ojModX.x[5],2)

1.88

In Year 3, 1.88 acres of Unit 2 should be cut down